File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        def remove_nodes(head):
            nonlocal max_val
            if(not head):
                return None

            head.next = remove_nodes(head.next)
            if(max_val > head.val):
                return head.next
            
            max_val = head.val
            return head
            
            
        
        max_val = -inf
        return remove_nodes(head)

File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNodes(head *ListNode) *ListNode {
  stack := list.New()
  for head != nil {
      for stack.Len() > 0 && 
      stack.Back().Value.(*ListNode).Val < head.Val{
          node := stack.Back()
          stack.Remove(node)
      }
      
      nxt := head.Next
      head.Next = nil
      stack.PushBack(head)
      head = nxt
  }
  
  head = &ListNode{
      Val: -1,
      Next: nil,
  }
  
  curr := head
  for stack.Len() > 0 {
      node := stack.Front()
      stack.Remove(node)
      curr.Next = node.Value.(*ListNode)
      curr = curr.Next
  }
  return head.Next
}